### ISSUE TITLE
Create workbooks for the katas/tutorials, part 1

### ISSUE THREAD

**Issue by tcNickolas** [microsoft/QuantumKatas#208]
The katas offer a set of programming exercises on quantum computing and reference solutions to them, but we often hear that it would be helpful to have the solutions explained, with the logic steps necessary to arrive from the problem description to the code spelled out. It is especially important for people who go through the katas on their own, without a study group to support them.

To address this, we came up with the idea of "workbooks" - explanations of the solutions to the tasks offered in the katas/tutorials, written in Jupyter Notebook format. You can see the workbook for Superposition kata [here](https://github.com/microsoft/QuantumKatas/blob/master/Superposition/Workbook_Superposition.ipynb) - it is still a work in progress, but it already shows the idea of the workbook, the kind of the content it should have and the way it accompanies the kata.

There are several other katas/tutorials that are a great fit for this kind of learning material:

- [x] Complex arithmetic (classical)
- [x] Linear algebra (classical)
- [x] Single-qubit gates
- [x] Multi-qubit systems
- [x] Multi-qubit gates
- [x] Superposition
- [x] Measurements
- [x] Joint measurements
- [x] SolveSATWithGrover
- [x] GraphColoring
- [x] CHSH game
- [x] GHZ game
- [x] Unitary patterns

Other katas might fit this scheme a bit worse - for example, Superdense Coding kata follows the structure of superdense coding algorithm, so a tutorial on the end-to-end algorithm would work better than independent explanations of individual tasks.

**Comment by vivanwin** [comment]
If there is still help needed I could start working on the first two.

**Comment by tcNickolas** [comment]
Nobody is working on those yet, help will be most definitely appreciated!

**Comment by eloiseespel** [comment]
I would love to help too, I'll start working on Measurements!

**Comment by tcNickolas** [comment]
Sounds great!

@vivanwin and @eloiseespel Please don't feel obliged to do the whole workbook at once (it's a lot of work!) - you can do the first couple of tasks and send them for review to get early feedback. Thank you again for volunteering!

**Comment by tcNickolas** [comment]
@eloiseespel How are you doing with the Measurements workbook? Do you need any help? Please don't hesitate to reach out with questions! And if you ran out of time after doing just a few tasks, feel free to send them in - Measurements is a massive batch of problems, and doing it in one go is an ambitious endeavor!

**Comment by eloiseespel** [comment]
@tcNickolas I completed the first 5 tasks early January I thought it merged correctly put since you have not received them I assume it didn't. I will merge again later today. Since then I have finished the first 10 tasks so I will send that, I would love feedback on those once I figure out the merging!

**Comment by desireevl** [comment]
Hey @tcNickolas I'd be happy to work on the single qubit gate workbook

**Comment by tcNickolas** [comment]
@desireevl Sounds great! You can see Superposition workbook as a great example of the kind of explanations we're looking for (although BasicGates ones will be much simpler!) and formatting.

**Comment by vivanwin** [comment]
Hi @tcNickolas, I can have a look at the multi-qubit gates.

**Comment by desireevl** [comment]
I am having issues with getting the exercises in `SingleQubitGates.ipynb` to run. I am using version:
Microsoft.Quantum.Standard::0.10.1911.1607
Microsoft.Quantum.Katas::0.10.1911.1607

I am able to run the demo however get the following error for Exercise 1:
```
%kata T1_ApplyY_Test

operation ApplyY (q : Qubit) : Unit is Adj+Ctl {
    Y(q);
}
```

`/snippet_.qs(1,85): error QS3001: Syntax does not match any known patterns.`

@tcNickolas would you have any insight as to what might be the issue?

**Comment by vivanwin** [comment]
Hi @desireevl,

I just tried to reproduce your error, for me this only happens when I have not run the first cell `%package Microsoft.Quantum.Katas::0.10.1911.1607`.
The issue is fixed when you run that cell and try to run your code again. Note that if you use the Binder  Notebook the kernel shuts down every ~10 minutes so you have to rerun that first cell again after the kernel is up again.

Hopefully this helps.

**Comment by vivanwin** [comment]
Hi,
I can pickup the Multi-quibt systems workbook.

**Comment by tcNickolas** [comment]
@vivanwin Sounds great, thank you!

**Comment by tcNickolas** [comment]
@desireevl Are you still working on the single qubit gate workbook? Please let us know if you ran into any issues.

**Comment by desireevl** [comment]
@tcNickolas Hi yes, sorry for the delay, been a bit hectic with the situation at the moment. Will be finished shortly!

**Comment by tcNickolas** [comment]
@desireevl I understand, no rush - just wanted to make sure you're not blocked on something I can help you with.

**Comment by RankJay** [comment]
@tcNickolas I would like to pick up the Measurements (tutorials) if that is available.

**Comment by manav28** [comment]
@tcNickolas I would like to take up Joint Measurements if no one has started already.

**Comment by tcNickolas** [comment]
@manav28 Sounds great, thank you!

**Comment by jainvasu631** [comment]
If no one has started working on it, I would like to call dibs on Unitary Patterns.

**Comment by tcNickolas** [comment]
@jainvasu631 Sounds great, thank you!

**Comment by tcNickolas** [comment]
@jainvasu631 The problems in UnitaryPatterns kata were explained in the editorials for the past contest, for the [warmup round](https://assets.codeforces.com/rounds/1115/warmup-editorial.pdf) and for the [main contest](https://assets.codeforces.com/rounds/1116/contest-editorial.pdf) - here are the LaTeX sources I have for them, in case you want to have a nice starting point for all the formulas and matrices. (They are saved as txt so as to allow uploading them to GitHub) 

[contest-editorial.txt](https://github.com/microsoft/QuantumKatas/files/4777109/contest-editorial.txt)
[warmup-editorial.txt](https://github.com/microsoft/QuantumKatas/files/4777112/warmup-editorial.txt)

Hope that helps!

**Comment by ArthurConmy** [comment]
@manav28 are you still interested in writing the workbook for joint measurements? If not, I am willing to work on it. Regardless, I have a short solution for task 5:

`%kata T05_DifferentBasis_Test

operation DifferentBasis (qs : Qubit[]) : Int {
    // ...
    
    CNOT(qs[1],qs[0]);
    H(qs[1]);
    
    mutable ans=0;
    
    if(M(qs[1]) == One)
    {
        set ans=1;
    }
    
    H(qs[1]);
    CNOT(qs[1],qs[0]); // reset to original state
    
    return ans;
}`

**Comment by manav28** [comment]
@ArthurConmy I am almost done with it. I just have task 6 pending. I believe task 5 has a clean one liner and I've written some of the math that goes into it
```
open Microsoft.Quantum.Measurement;

operation DifferentBasis (qs : Qubit[]) : Int {
    // ...
    return Measure([PauliX, PauliX], qs) == Zero ? 0 | 1;
}
```

**Comment by jainvasu631** [comment]
@tcNickolas I noticed that Grover's Algorithm though not on this list also doesn't have a Workbook. 
Although the problems in it are quite simple. So do you think it needs a Workbook. If yes then I wouldn't mind working on it.

**Comment by tcNickolas** [comment]
@jainvasu631 I'm a bit on the fence about Grover's algorithm kata - on one hand, if there are tasks, it's good to have a workbook for them; but on the other hand, the tasks follow the algorithm exactly, there's less space for inventing the solution than in some of the other katas (that's why I didn't include this kata in the first wave of workbooks to be developed). I would say it's worth having if you're willing to create it. Thank you!

(Sorry for the delay replying on this and on the PRs, but I'm returning to the open source work now so should be a bit more responsive :-))

**Comment by tcNickolas** [comment]
@desireevl Hi. Are you still working on the single qubit gate workbook? Please let me know - if you have it in some partial state but don't have time to finish it, you're welcome to send PR with this partial, and somebody else will pick it up to finish. (We have an internal Hackathon approaching so we'll have extra volunteers looking for something to do, and this workbook is a good candidate :-))

**Comment by desireevl** [comment]
Hi @tcNickolas, I am so sorry about this. I won't be able to work on this and have unfortunately lost my progress  for what I have done

**Comment by tcNickolas** [comment]
@desireevl No worries at all, I know how it can be. Thank you for letting me know!

**Comment by Lukious** [comment]
If there are no applicants for GraphColoring yet, I'd like to work on it. Would that be okay?

**Comment by tcNickolas** [comment]
@Lukious Absolutely, thank you!

**Comment by jainvasu631** [comment]
@tcNickolas I wouldn't mind taking over Single Qubit Workbook from @desireevl . Its relatively simpler and I agree the Grover's algorithm Kata would be too directly related to the algorithm so maybe its not a priority right now.

**Comment by tcNickolas** [comment]
@jainvasu631 Sounds good!

**Comment by tcNickolas** [comment]
@ArthurConmy I finally finished review of JointMeasurements workbook and merged it to master. Looking at [your proposed solution to task 5](https://github.com/microsoft/QuantumKatas/issues/208#issuecomment-644909162), it seems to follow quite a different approach, so I think there is value in adding it (together with the math behind it) after the existing solution, as an alternate solution (the way we've done in some Superposition tasks). Let me know if you'd like to do that!

**Comment by BrokenDuck** [comment]
@Lukious How far are you with the graph coloring workbook ? I can help if you want !

**Comment by BrokenDuck** [comment]
@tcNickolas Can I work on the graph coloring workbook then ?

**Comment by tcNickolas** [comment]
@BrokenDuck Sure, sounds good :-) 

(Apologies for the delay replying - I was trying to reach somebody who has expressed interest in the same some time ago, but with no success, so I'm assuming they're not interested any longer.)

**Comment by jainvasu631** [comment]
@tcNickolas 
I was recently going through the QFT Kata Notebook after Section III was added. I felt that while Section I is maybe tutorial-ish,
Section II and especially Section III need a Workbook. None currently exists. The list should be updated with the QFT as well.

PS: I notice that its almost been a year since this Issue was opened (in the last Hacktoberfest). We have almost completed it and it would be extremely cathartic if we could/would be to close it on October 28 (exactly one year after it was created). However, maybe we should expand the list to include new candidate Katas/Tutorial which are ripe and in need of a Workbook.

**Comment by tcNickolas** [comment]
@jainvasu631 
Agreed on both counts - I was hesitant to add new workbooks here, since this would delay the joy of completing the issue for everyone who worked so hard on the first batch of the workbooks :-)
I'm thinking we can file a new issue, "Create workbooks, batch 2", so that we can both complete this one and have more workbooks to work on.

Which reminds me... @BrokenDuck, how are you doing with the workbook for the graph coloring kata? Let us know if you need any help!

**Comment by BrokenDuck** [comment]
@tcNickolas I am working on it, but school has been quite ruthless (4h of French, Maths and Physics exam), so I am a bit late. I would greatly appreciate some help on part 1, I am more interested in part 2. Do I just commit the partially finished workbook, or are there extra steps ?

**Comment by tcNickolas** [comment]
@BrokenDuck I think it's fine to send a pull request for a partially finished workbook - we've done that with Superposition, which took a join effort of 4 or 5 people and probably around 10 PRs.

**Comment by jainvasu631** [comment]
@BrokenDuck  I would love to help out on GraphColoring's Workbook. If you can maybe send a draft PR or share whatever you have worked on, I would be glad to help out.

**Comment by tcNickolas** [comment]
@BrokenDuck @jainvasu631 Here's the [history of commits on Superposition workbook](https://github.com/microsoft/QuantumKatas/commits/main/Superposition/Workbook_Superposition.ipynb), as an example of what we did when different people work on different sections.

**Comment by jainvasu631** [comment]
@tcNickolas @BrokenDuck  
Perhaps we could commit Part 1 and Part 2 independently. I am not sure how much progress @BrokenDuck has made on Part 1 but I could build on that and finish it before end of October.

@BrokenDuck How does that sound?

**Comment by BrokenDuck** [comment]
I didn't really start part 1, so you could do that part and I can continue on part 2 with the oracle

**Comment by jainvasu631** [comment]
@tcNickolas @BrokenDuck 
I have written the first draft of the Workbook Part 1. The Part of the Workbook can then be merged into this. I have right now created an empty coming soon Solution cells for the tasks in Part 2. Later on these can be filled in.

**Comment by BrokenDuck** [comment]
@tcNickolas I finished the graph coloring workbook, I sent a push request.

**Comment by tcNickolas** [comment]
With the GraphColoring workbook finished, I officially declare this issue complete! Thank you everybody who contributed to it!

I'll go through the remaining katas and put together a plan for the second batch of the workbooks soon.
